common/lib/dependabot/shared_helpers.rb

```diff
@@ -280,10 +280,11 @@ def self.reset_global_git_config(backup_path)
       FileUtils.mv(backup_path, GIT_CONFIG_GLOBAL_PATH)
     end
 
-    def self.run_shell_command(command, allow_unsafe_shell_command: false)
+    def self.run_shell_command(command, allow_unsafe_shell_command: false, env: nil)
       start = Time.now
       cmd = allow_unsafe_shell_command ? command : escape_command(command)
-      stdout, process = Open3.capture2e(cmd)
+      env_cmd = [env, cmd].compact
+      stdout, process = Open3.capture2e(*env_cmd)
```

Comment:
Could we rephrase this to make the intent of this a bit clearer? It took me a few minutes to understand what was happening here.

At first, I thought _"`env` is something like `TEST_ENV=prefix`, so this concatenates the original command"_. But then I realized that `env` is a hash, and had to consult the [docs for `Process#spawn`](https://docs.ruby-lang.org/en/2.0.0/Process.html#method-c-spawn) to understand how `compact` is called here to mitigate the awkwardness of `#spawn` taking `env` as an optional first argument.

Would something like this work instead?

```ruby
def self.run_shell_command(command, allow_unsafe_shell_command: false, env: {})
      start = Time.now
      cmd = allow_unsafe_shell_command ? command : escape_command(command)
      stdout, process = Open3.capture2e(env, cmd)
```

Reply:
I think using a default of `env: {}` in the method signature might clear this up as well. We don't have a guarantee against `nil` being passed in, but we catch the cases where it is not supplied 🙂

Reply:
👍 the optional first argument has thrown me off a few times as well, it's such a weird API

Reply:
Love these ideas! Updated in https://github.com/dependabot/dependabot-core/pull/4434/commits/16395d30cedb5554090873ece20d99321cea4fc7.

There is another usage of compact at https://github.com/dependabot/dependabot-core/blob/ebe338ed2bd1d09b2c86510f9e44cc0cf642e18b/common/lib/dependabot/shared_helpers.rb#L93-L94 but I'm going to leave that for a future PR.